### PR TITLE
Issue 676 | Persistent storage: pvc for ray nodes

### LINKED ISSUE #676: Gateway: add PVC to cluster
### What is the expected enhancement?

As https://github.com/Qiskit-Extensions/quantum-serverless/pull/670 got merged we can do use this configmap to attach PVC for persistent storage to ray nodes. 

Part of #463.

### PR DESCRIPTION
**Pull request by IceKhan13**
### Summary

Persistent storage: pvc for ray nodes

### Details and comments

Closes #676

### CODE REVIEW

**Comment by IceKhan13** [comment]
For some reason `/data` folder is not mounted properly and always empty 🤔 
I recheck volume and subpath is created properly and I have folder to write to. But whenever I write from ray node to `/data` folder it is not saved

**Comment by akihikokuroda** [comment]
@IceKhan13 The volumeMount is in the side car container(ray-head-logs) instead of the main container(ray-head).

**Comment by IceKhan13** [comment]
hahah 😄 omg, thats true

**Comment by IceKhan13** [comment]
Forgot to rebuild gateway image. Looks like everything is working and subpath mounting part of volume correctly. 
Only thing remaining is setup correct permissions to write to volumes as gateway has user 1001 and ray is using 1000

**Comment by IceKhan13** [comment]
Now each user has it's own folder where it can write and read data from. Location is `/data` which is mounted for each user for their ray node.

**Review by psschwei — COMMENTED** [review]
for me, using `platform: local` on kind, the PVC is still owned by root
edit: seems we went with `kind` instead of `local` for the tag... let me retry
edit2: `platform: kind` mounts the PVC fine, but now running a program throws a `ConnectionError: Failed to connect to Ray at address: http://user-9a96cfa7-head-svc:8265.` error in the scheduler

**Comment by IceKhan13** [comment]
@psschwei can you post logs from scheduler?

**Comment by psschwei** [comment]
Scheduler logs:

---

Using selector: EpollSelector
4 free cluster slots.
Head node http://user-0bfe3464-head-svc:8265/ is not ready yet.
Using API server address http://user-0bfe3464-head-svc:8265/.
Uploading package gcs://_ray_pkg_6fcca69c75b5b5c8.zip.
Creating a file package for local directory '/usr/src/app/media/tmp/8c81d2fd-b65f-409f-95d7-2c75f905d458'.
Rewrote runtime_env `pip` field from [] to None.
Traceback (most recent call last):
  File "/usr/local/lib/python3.9/site-packages/urllib3/connection.py", line 174, in _new_conn
    conn = connection.create_connection(
  File "/usr/local/lib/python3.9/site-packages/urllib3/util/connection.py", line 72, in create_connection
    for res in socket.getaddrinfo(host, port, family, socket.SOCK_STREAM):
  File "/usr/local/lib/python3.9/socket.py", line 954, in getaddrinfo
    for res in _socket.getaddrinfo(host, port, family, type, proto, flags):
socket.gaierror: [Errno -2] Name or service not known

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/local/lib/python3.9/site-packages/urllib3/connectionpool.py", line 714, in urlopen
    httplib_response = self._make_request(
  File "/usr/local/lib/python3.9/site-packages/urllib3/connectionpool.py", line 415, in _make_request
    conn.request(method, url, **httplib_request_kw)
  File "/usr/local/lib/python3.9/site-packages/urllib3/connection.py", line 244, in request
    super(HTTPConnection, self).request(method, url, body=body, headers=headers)
  File "/usr/local/lib/python3.9/http/client.py", line 1285, in request
    self._send_request(method, url, body, headers, encode_chunked)
  File "/usr/local/lib/python3.9/http/client.py", line 1331, in _send_request
    self.endheaders(body, encode_chunked=encode_chunked)
  File "/usr/local/lib/python3.9/http/client.py", line 1280, in endheaders
    self._send_output(message_body, encode_chunked=encode_chunked)
  File "/usr/local/lib/python3.9/http/client.py", line 1040, in _send_output
    self.send(msg)
  File "/usr/local/lib/python3.9/http/client.py", line 980, in send
    self.connect()
  File "/usr/local/lib/python3.9/site-packages/urllib3/connection.py", line 205, in connect
    conn = self._new_conn()
  File "/usr/local/lib/python3.9/site-packages/urllib3/connection.py", line 186, in _new_conn
    raise NewConnectionError(
urllib3.exceptions.NewConnectionError: <urllib3.connection.HTTPConnection object at 0x7f816cb59400>: Failed to establish a new connection: [Errno -2] Name or service not known

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/local/lib/python3.9/site-packages/requests/adapters.py", line 486, in send
    resp = conn.urlopen(
  File "/usr/local/lib/python3.9/site-packages/urllib3/connectionpool.py", line 798, in urlopen
    retries = retries.increment(
  File "/usr/local/lib/python3.9/site-packages/urllib3/util/retry.py", line 592, in increment
    raise MaxRetryError(_pool, url, error or ResponseError(cause))
urllib3.exceptions.MaxRetryError: HTTPConnectionPool(host='user-0bfe3464-head-svc', port=8265): Max retries exceeded with url: /api/jobs/ (Caused by NewConnectionError('<urllib3.connection.HTTPConnection object at 0x7f816cb59400>: Failed to establish a new connection: [Errno -2] Name or service not known'))

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/src/app/manage.py", line 22, in <module>
    main()
  File "/usr/src/app/manage.py", line 18, in main
    execute_from_command_line(sys.argv)
  File "/usr/local/lib/python3.9/site-packages/django/core/management/__init__.py", line 442, in execute_from_command_line
    utility.execute()
  File "/usr/local/lib/python3.9/site-packages/django/core/management/__init__.py", line 436, in execute
    self.fetch_command(subcommand).run_from_argv(self.argv)
  File "/usr/local/lib/python3.9/site-packages/django/core/management/base.py", line 412, in run_from_argv
    self.execute(*args, **cmd_options)
  File "/usr/local/lib/python3.9/site-packages/django/core/management/base.py", line 458, in execute
    output = self.handle(*args, **options)
  File "/usr/src/app/api/management/commands/schedule_queued_jobs.py", line 56, in handle
    job = execute_job(job)
  File "/usr/src/app/api/schedule.py", line 61, in execute_job
    job = submit_ray_job(job)
  File "/usr/src/app/api/ray.py", line 43, in submit_ray_job
    ray_job_id = ray_client.submit_job(
  File "/usr/local/lib/python3.9/site-packages/ray/dashboard/modules/job/sdk.py", line 226, in submit_job
    r = self._do_request("POST", "/api/jobs/", json_data=json_data)
  File "/usr/local/lib/python3.9/site-packages/ray/dashboard/modules/dashboard_sdk.py", line 303, in _do_request
    return requests.request(
  File "/usr/local/lib/python3.9/site-packages/requests/api.py", line 59, in request
    return session.request(method=method, url=url, **kwargs)
  File "/usr/local/lib/python3.9/site-packages/requests/sessions.py", line 589, in request
    resp = self.send(prep, **send_kwargs)
  File "/usr/local/lib/python3.9/site-packages/requests/sessions.py", line 703, in send
    r = adapter.send(request, **kwargs)
  File "/usr/local/lib/python3.9/site-packages/requests/adapters.py", line 519, in send
    raise ConnectionError(e, request=request)
requests.exceptions.ConnectionError: HTTPConnectionPool(host='user-0bfe3464-head-svc', port=8265): Max retries exceeded with url: /api/jobs/ (Caused by NewConnectionError('<urllib3.connection.HTTPConnection object at 0x7f816cb59400>: Failed to establish a new connection: [Errno -2] Name or service not known'))

---

**Comment by IceKhan13** [comment]
hm, seems like it has nothing to do with this PR 🤔 
nothing has changed on connection to ray node side. Only thing was added to ray node is folder and that is it

**Comment by IceKhan13** [comment]
did ray node builded correctly?

**Comment by IceKhan13** [comment]
are images pulled to cluster?

**Comment by psschwei** [comment]
yes.

weirdly, on the runs that fail, the containers get stopped 14 seconds after they start:

```
Events:
  Type    Reason     Age   From               Message
  ----    ------     ----  ----               -------
  Normal  Scheduled  36s   default-scheduler  Successfully assigned default/user-5e9e1a0e-head-n4fbq to kind-control-plane
  Normal  Pulled     35s   kubelet            Container image "ray:test" already present on machine
  Normal  Created    35s   kubelet            Created container ray-head
  Normal  Started    35s   kubelet            Started container ray-head
  Normal  Pulled     35s   kubelet            Container image "fluent/fluent-bit:1.9.10" already present on machine
  Normal  Created    35s   kubelet            Created container ray-head-logs
  Normal  Started    35s   kubelet            Started container ray-head-logs
  Normal  Killing    21s   kubelet            Stopping container ray-head
  Normal  Killing    21s   kubelet            Stopping container ray-head-logs
```

whereas the one that succeeded kept running for >30s:

```
Events:
  Type    Reason     Age   From               Message
  ----    ------     ----  ----               -------
  Normal  Scheduled  37s   default-scheduler  Successfully assigned default/user-fff4c468-head-hc8cw to kind-control-plane
  Normal  Pulled     36s   kubelet            Container image "ray:test" already present on machine
  Normal  Created    36s   kubelet            Created container ray-head
  Normal  Started    36s   kubelet            Started container ray-head
  Normal  Pulled     36s   kubelet            Container image "fluent/fluent-bit:1.9.10" already present on machine
  Normal  Created    36s   kubelet            Created container ray-head-logs
  Normal  Started    35s   kubelet            Started container ray-head-logs
```

since the pods get killed rapidly, takes a bit of finger jujitsu to catch logs / events from them realtime

**Comment by IceKhan13** [comment]
I have no idea why is this happening :) 
and I cannot reproduce it too =(

**Comment by psschwei** [comment]
> if you do not update ray image? is it working?

when I built off of `main`, that was working... haven't tried with `latest` or the a versioned tag

**Comment by IceKhan13** [comment]
are those jobs returned with failed status?

**Comment by psschwei** [comment]
no, the job status is "QUEUED"

**Comment by IceKhan13** [comment]
> user-0073e351-head-qfpdh ray-head-logs [2023/06/14 17:51:46] [engine] caught signal (SIGTERM)

oh, looks like that is the one

**Comment by IceKhan13** [comment]
> no, the job status is "QUEUED"

ok, scheduler just trying to reschedule this job again, that is why you have new cluster. 
If cluster is created, job is launched and then something happened, then it will be in failed status

**Comment by psschwei** [comment]
these are all the logs from one set of ray pods:

---

+ user-be601614-worker-smallworkergroup-kpzcg › wait-gcs-ready
+ user-be601614-head-hn9p2 › ray-head-logs
+ user-be601614-head-hn9p2 › ray-head
user-be601614-head-hn9p2 ray-head-logs Fluent Bit v1.9.10
user-be601614-head-hn9p2 ray-head-logs * Copyright (C) 2015-2022 The Fluent Bit Authors
user-be601614-head-hn9p2 ray-head-logs * Fluent Bit is a CNCF sub-project under the umbrella of Fluentd
user-be601614-head-hn9p2 ray-head-logs * https://fluentbit.io
user-be601614-head-hn9p2 ray-head-logs 
user-be601614-head-hn9p2 ray-head-logs [2023/06/14 17:50:06] [ info] [fluent bit] version=1.9.10, commit=760956f50c, pid=1
user-be601614-head-hn9p2 ray-head-logs [2023/06/14 17:50:06] [ info] [storage] version=1.3.0, type=memory-only, sync=normal, checksum=disabled, max_chunks_up=128
user-be601614-head-hn9p2 ray-head-logs [2023/06/14 17:50:06] [ info] [cmetrics] version=0.3.7
user-be601614-head-hn9p2 ray-head-logs [2023/06/14 17:50:06] [error] [input:tail:tail.0] read error, check permissions: /tmp/ray/session_latest/logs/job-driver-raysubmit_*
user-be601614-head-hn9p2 ray-head-logs [2023/06/14 17:50:06] [ warn] [input:tail:tail.0] error scanning path: /tmp/ray/session_latest/logs/job-driver-raysubmit_*
user-be601614-head-hn9p2 ray-head-logs [2023/06/14 17:50:06] [ info] [sp] stream processor started
user-be601614-head-hn9p2 ray-head-logs [2023/06/14 17:50:06] [ info] [output:stdout:stdout.0] worker #0 started
- user-be601614-worker-smallworkergroup-kpzcg › wait-gcs-ready
user-be601614-head-hn9p2 ray-head 2023-06-14 10:50:07,418       INFO usage_lib.py:398 -- Usage stats collection is enabled by default without user confirmation because this terminal is detected to be non-interactive. To disable this, add `--disable-usage-stats` to the command that starts the cluster, or run the following command: `ray disable-usage-stats` before starting the cluster. See https://docs.ray.io/en/master/cluster/usage-stats.html for more details.
user-be601614-head-hn9p2 ray-head 2023-06-14 10:50:07,418       INFO scripts.py:710 -- Local node IP: 10.244.0.92
user-be601614-head-hn9p2 ray-head 2023-06-14 10:50:09,651       SUCC scripts.py:747 -- --------------------
user-be601614-head-hn9p2 ray-head 2023-06-14 10:50:09,652       SUCC scripts.py:748 -- Ray runtime started.
user-be601614-head-hn9p2 ray-head 2023-06-14 10:50:09,652       SUCC scripts.py:749 -- --------------------
user-be601614-head-hn9p2 ray-head 2023-06-14 10:50:09,652       INFO scripts.py:751 -- Next steps
user-be601614-head-hn9p2 ray-head 2023-06-14 10:50:09,652       INFO scripts.py:754 -- To add another node to this Ray cluster, run
user-be601614-head-hn9p2 ray-head 2023-06-14 10:50:09,652       INFO scripts.py:757 --   ray start --address='10.244.0.92:6379'
user-be601614-head-hn9p2 ray-head 2023-06-14 10:50:09,652       INFO scripts.py:766 -- To connect to this Ray cluster:
user-be601614-head-hn9p2 ray-head 2023-06-14 10:50:09,652       INFO scripts.py:768 -- import ray
user-be601614-head-hn9p2 ray-head 2023-06-14 10:50:09,652       INFO scripts.py:769 -- ray.init()
user-be601614-head-hn9p2 ray-head 2023-06-14 10:50:09,652       INFO scripts.py:781 -- To submit a Ray job using the Ray Jobs CLI:
user-be601614-head-hn9p2 ray-head 2023-06-14 10:50:09,652       INFO scripts.py:782 --   RAY_ADDRESS='http://10.244.0.92:8265' ray job submit --working-dir . -- python my_script.py
user-be601614-head-hn9p2 ray-head 2023-06-14 10:50:09,652       INFO scripts.py:791 -- See https://docs.ray.io/en/latest/cluster/running-applications/job-submission/index.html 
user-be601614-head-hn9p2 ray-head 2023-06-14 10:50:09,652       INFO scripts.py:795 -- for more information on submitting Ray jobs to the Ray cluster.
user-be601614-head-hn9p2 ray-head 2023-06-14 10:50:09,652       INFO scripts.py:800 -- To terminate the Ray runtime, run
user-be601614-head-hn9p2 ray-head 2023-06-14 10:50:09,652       INFO scripts.py:801 --   ray stop
user-be601614-head-hn9p2 ray-head 2023-06-14 10:50:09,652       INFO scripts.py:804 -- To view the status of the cluster, use
user-be601614-head-hn9p2 ray-head 2023-06-14 10:50:09,652       INFO scripts.py:805 --   ray status
user-be601614-head-hn9p2 ray-head 2023-06-14 10:50:09,652       INFO scripts.py:809 -- To monitor and debug Ray, view the dashboard at 
user-be601614-head-hn9p2 ray-head 2023-06-14 10:50:09,652       INFO scripts.py:810 --   10.244.0.92:8265
user-be601614-head-hn9p2 ray-head 2023-06-14 10:50:09,652       INFO scripts.py:817 -- If connection to the dashboard fails, check your firewall settings and network configuration.
user-be601614-head-hn9p2 ray-head 2023-06-14 10:50:09,652       INFO scripts.py:917 -- --block
user-be601614-head-hn9p2 ray-head 2023-06-14 10:50:09,653       INFO scripts.py:918 -- This command will now block forever until terminated by a signal.
user-be601614-head-hn9p2 ray-head 2023-06-14 10:50:09,653       INFO scripts.py:921 -- Running subprocesses are monitored and a message will be printed if any of them terminate unexpectedly. Subprocesses exit with SIGTERM will be treated as graceful, thus NOT reported.
+ user-be601614-worker-smallworkergroup-kpzcg › ray-worker
user-be601614-worker-smallworkergroup-kpzcg ray-worker 2023-06-14 10:50:10,401  INFO scripts.py:892 -- Local node IP: 127.0.0.1
user-be601614-worker-smallworkergroup-kpzcg ray-worker 2023-06-14 10:50:15,990  SUCC scripts.py:904 -- --------------------
user-be601614-worker-smallworkergroup-kpzcg ray-worker 2023-06-14 10:50:15,991  SUCC scripts.py:905 -- Ray runtime started.
user-be601614-worker-smallworkergroup-kpzcg ray-worker 2023-06-14 10:50:15,991  SUCC scripts.py:906 -- --------------------
user-be601614-worker-smallworkergroup-kpzcg ray-worker 2023-06-14 10:50:15,991  INFO scripts.py:908 -- To terminate the Ray runtime, run
user-be601614-worker-smallworkergroup-kpzcg ray-worker 2023-06-14 10:50:15,991  INFO scripts.py:909 --   ray stop
user-be601614-worker-smallworkergroup-kpzcg ray-worker 2023-06-14 10:50:15,991  INFO scripts.py:917 -- --block
user-be601614-worker-smallworkergroup-kpzcg ray-worker 2023-06-14 10:50:15,991  INFO scripts.py:918 -- This command will now block forever until terminated by a signal.
user-be601614-worker-smallworkergroup-kpzcg ray-worker 2023-06-14 10:50:15,991  INFO scripts.py:921 -- Running subprocesses are monitored and a message will be printed if any of them terminate unexpectedly. Subprocesses exit with SIGTERM will be treated as graceful, thus NOT reported.
user-be601614-head-hn9p2 ray-head-logs [2023/06/14 17:50:20] [engine] caught signal (SIGTERM)
user-be601614-head-hn9p2 ray-head-logs [2023/06/14 17:50:20] [ info] [input] pausing tail.0
user-be601614-head-hn9p2 ray-head-logs [2023/06/14 17:50:20] [ warn] [engine] service will shutdown in max 5 seconds
user-be601614-head-hn9p2 ray-head-logs [2023/06/14 17:50:21] [ info] [engine] service has stopped (0 pending tasks)
user-be601614-head-hn9p2 ray-head-logs [2023/06/14 17:50:21] [ info] [output:stdout:stdout.0] thread worker #0 stopping...
user-be601614-head-hn9p2 ray-head-logs [2023/06/14 17:50:21] [ info] [output:stdout:stdout.0] thread worker #0 stopped
- user-be601614-head-hn9p2 › ray-head
- user-be601614-head-hn9p2 › ray-head-logs

**Comment by IceKhan13** [comment]
> user-be601614-head-hn9p2 ray-head 2023-06-14 10:50:09,652 INFO scripts.py:810 -- 10.244.0.92:8265
user-be601614-head-hn9p2 ray-head 2023-06-14 10:50:09,652 INFO scripts.py:817 -- If connection to the dashboard fails, check your firewall settings and network configuration.
user-be601614-head-hn9p2 ray-head 2023-06-14 10:50:09,652 INFO scripts.py:917 -- --block
user-be601614-head-hn9p2 ray-head 2023-06-14 10:50:09,653 INFO scripts.py:918 -- This command will now block forever until terminated by a signal.
user-be601614-head-hn9p2 ray-head 2023-06-14 10:50:09,653 INFO scripts.py:921 -- Running subprocesses are monitored and a message will be printed if any of them terminate unexpectedly. Subprocesses exit with SIGTERM will be treated as graceful, thus NOT reported.

hmmm

File: gateway/api/ray.py [hunk 1]
```diff
@@ -77,7 +77,10 @@ def create_ray_cluster(
     cluster_name = cluster_name or f"{user.username}-{str(uuid.uuid4())[:8]}"
     if not cluster_data:
         cluster = get_template("rayclustertemplate.yaml")
-        cluster_data = yaml.safe_load(cluster.render({"cluster_name": cluster_name}))
+        manifest = cluster.render(
+            {"cluster_name": cluster_name, "user_id": user.username}
```
Comment (psschwei) [inline gateway/api/ray.py:81]: thinking out loud... is there any chance that `user` could be `None` here?

Comment (IceKhan13) [reply gateway/api/ray.py:81]: only if job has no user associated to it, which is not possible as all jobs are authored by authentication, which is only way to access endpoints

Comment (psschwei) [reply gateway/api/ray.py:81]: nah, this was a dead end

**Comment by psschwei** [comment]
a bit more info:

when this is failing, port 8265 isn't open for connections:

```
oot@paultest:/# nmap -Pn -p 8265 user-963d0ad5-head-svc
Starting Nmap 7.70 ( https://nmap.org ) at 2023-06-14 18:44 UTC
Nmap scan report for user-963d0ad5-head-svc (10.96.105.242)
Host is up (0.014s latency).

PORT     STATE    SERVICE
8265/tcp filtered unknown

Nmap done: 1 IP address (1 host up) scanned in 0.52 seconds
```

however, once the port is open, then everything works as expected:

```
root@paultest:/# nmap -Pn -p 8265 user-b82c9c97-head-svc
Starting Nmap 7.70 ( https://nmap.org ) at 2023-06-14 18:46 UTC
Nmap scan report for user-b82c9c97-head-svc (10.96.25.175)
Host is up (0.000042s latency).
rDNS record for 10.96.25.175: user-b82c9c97-head-svc.default.svc.cluster.local

PORT     STATE SERVICE
8265/tcp open  unknown

Nmap done: 1 IP address (1 host up) scanned in 0.27 seconds
```

now as to why the port is open sometimes and not at others... :shrug:

**Comment by IceKhan13** [comment]
> now as to why the port is open sometimes and not at others... 🤷

and how is it related to changes in this PR

**Comment by IceKhan13** [comment]
is your postgres volume cleared? 🤔

**Comment by psschwei** [comment]
> is your postgres volume cleared? thinking

not sure ... deleted all volumes and will try again

**Comment by IceKhan13** [comment]
wtf

**Comment by psschwei** [comment]
ok, after banging my head against the wall for most of the night on this, decided to try this on minikube and.... it works fine 

image

**Comment by psschwei** [comment]
Has anyone else got it working on kind? If it's just me, then I can use minikube and we can merge

If it's failing for others on kind (but working on other things like minikube / rancher ), then it could just be an issue with how kind handles PVCs and we can merge

For other cases, we should probably look into it a bit more first

**Comment by pacomf** [comment]
If it is working on docker and kubernetes (at all), we can consider merge it, and then we can analyze why it is failing through kind (we can open a new issue with the logs to analyze later). It can be great fix the kind installation, but maybe to avoid block other things, we can investigate in the future open an issue

**Comment by psschwei** [comment]
My only worry with merging now is that if it _is_ a bug somewhere, it'll get harder to find the bigger the codebase gets

**Comment by psschwei** [comment]
That said, I'm not sure if it's a bug in kind or our codebase, so curious how others have run it successfully

**Comment by IceKhan13** [comment]
90% sure it's kind configuration somewhere. It's almost impossible to get SEGKILL otherwise :) 

we can have a switch not to kill clusters, if we need it.

**Comment by akihikokuroda** [comment]
I'm trying to setup this quickly.  Do I need to rebuild gateway and ray node images?

**Comment by IceKhan13** [comment]
@akihikokuroda yes

**Comment by akihikokuroda** [comment]
I got an error.  Here is the logs in gateway.  I might not set up correctly.
```
Starting new HTTP connection (1): keycloak:31059
http://keycloak:31059 "POST /realms/quantumserverless/protocol/openid-connect/token/ HTTP/1.1" 200 3286
Starting new HTTP connection (1): gateway:8000
Starting new HTTP connection (1): keycloak:31059
http://keycloak:31059 "GET /realms/quantumserverless/protocol/openid-connect/userinfo HTTP/1.1" 200 133
http://gateway:8000 "POST /dj-rest-auth/keycloak/ HTTP/1.1" 200 356
Internal Server Error: /api/v1/programs/run/
Traceback (most recent call last):
  File "/usr/local/lib/python3.9/site-packages/django/core/handlers/exception.py", line 55, in inner
    response = get_response(request)
  File "/usr/local/lib/python3.9/site-packages/django/core/handlers/base.py", line 197, in _get_response
    response = wrapped_callback(request, *callback_args, **callback_kwargs)
  File "/usr/local/lib/python3.9/site-packages/django/views/decorators/csrf.py", line 56, in wrapper_view
    return view_func(*args, **kwargs)
  File "/usr/local/lib/python3.9/site-packages/rest_framework/viewsets.py", line 125, in view
    return self.dispatch(request, *args, **kwargs)
  File "/usr/local/lib/python3.9/site-packages/rest_framework/views.py", line 509, in dispatch
    response = self.handle_exception(exc)
  File "/usr/local/lib/python3.9/site-packages/rest_framework/views.py", line 469, in handle_exception
    self.raise_uncaught_exception(exc)
  File "/usr/local/lib/python3.9/site-packages/rest_framework/views.py", line 480, in raise_uncaught_exception
    raise exc
  File "/usr/local/lib/python3.9/site-packages/rest_framework/views.py", line 506, in dispatch
    response = handler(request, *args, **kwargs)
  File "/usr/src/app/api/views.py", line 61, in run
    program = save_program(serializer=serializer, request=request)
  File "/usr/src/app/api/schedule.py", line 29, in save_program
    program.save()
  File "/usr/local/lib/python3.9/site-packages/django/db/models/base.py", line 814, in save
    self.save_base(
  File "/usr/local/lib/python3.9/site-packages/django/db/models/base.py", line 877, in save_base
    updated = self._save_table(
  File "/usr/local/lib/python3.9/site-packages/django/db/models/base.py", line 1020, in _save_table
    results = self._do_insert(
  File "/usr/local/lib/python3.9/site-packages/django_prometheus/models.py", line 43, in _do_insert
    return super()._do_insert(*args, **kwargs)
  File "/usr/local/lib/python3.9/site-packages/django/db/models/base.py", line 1061, in _do_insert
    return manager._insert(
  File "/usr/local/lib/python3.9/site-packages/django/db/models/manager.py", line 87, in manager_method
    return getattr(self.get_queryset(), name)(*args, **kwargs)
  File "/usr/local/lib/python3.9/site-packages/django/db/models/query.py", line 1805, in _insert
    return query.get_compiler(using=using).execute_sql(returning_fields)
  File "/usr/local/lib/python3.9/site-packages/django/db/models/sql/compiler.py", line 1821, in execute_sql
    for sql, params in self.as_sql():
  File "/usr/local/lib/python3.9/site-packages/django/db/models/sql/compiler.py", line 1745, in as_sql
    value_rows = [
  File "/usr/local/lib/python3.9/site-packages/django/db/models/sql/compiler.py", line 1746, in <listcomp>
    [
  File "/usr/local/lib/python3.9/site-packages/django/db/models/sql/compiler.py", line 1747, in <listcomp>
    self.prepare_value(field, self.pre_save_val(field, obj))
  File "/usr/local/lib/python3.9/site-packages/django/db/models/sql/compiler.py", line 1695, in pre_save_val
    return field.pre_save(obj, add=True)
  File "/usr/local/lib/python3.9/site-packages/django/db/models/fields/files.py", line 317, in pre_save
    file.save(file.name, file.file, save=False)
  File "/usr/local/lib/python3.9/site-packages/django/db/models/fields/files.py", line 93, in save
    self.name = self.storage.save(name, content, max_length=self.field.max_length)
  File "/usr/local/lib/python3.9/site-packages/django/core/files/storage/base.py", line 38, in save
    name = self._save(name, content)
  File "/usr/local/lib/python3.9/site-packages/django/core/files/storage/filesystem.py", line 87, in _save
    os.makedirs(directory, exist_ok=True)
  File "/usr/local/lib/python3.9/os.py", line 225, in makedirs
    mkdir(name, mode)
PermissionError: [Errno 13] Permission denied: '/usr/src/app/media/user/f2ecc612-cf9b-4802-b5d5-b13e34fc5faf'
```

**Comment by psschwei** [comment]
Did you set `platform: kind` in the helm values? that's where I've experienced permission denied issues with the pvc

**Comment by IceKhan13** [comment]
make sure your volumes are empty and recreated.

```
/usr/src/app/media/user
```

folder probably owned by 1001 user from previous runs

**Comment by akihikokuroda** [comment]
@psschwei Oh, sorry.  This is with rancher desktop.  I wanted to make sure it works there first.

**Comment by psschwei** [comment]
After a bit more digging, I think I know what's going on here...

**Comment by IceKhan13** [comment]
cliffhanger :)

**Comment by psschwei** [comment]
The failures are happening in the [`submit_ray_job()`](https://github.com/Qiskit-Extensions/quantum-serverless/blob/a98fad67bebc5f98f47ba988eb19abb3ac10a0e7/gateway/api/ray.py#L25) function.

There's actually 2 places where things can fail:

1. [When we set up the `JobSubmissionClient`](https://github.com/Qiskit-Extensions/quantum-serverless/blob/a98fad67bebc5f98f47ba988eb19abb3ac10a0e7/gateway/api/ray.py#L34)

Seems like the JobSubmissionClient tries to make a connection to the Ray cluster, but if that fails it throws the `NewConnectionError` that I posted above.

2. [When using the ray client to submit the job](https://github.com/Qiskit-Extensions/quantum-serverless/blob/a98fad67bebc5f98f47ba988eb19abb3ac10a0e7/gateway/api/ray.py#L43)

I don't have an example of the logs for when this happens, so you'll have to take my word for it (it's ok, I'm not LeVar Burton).

In both cases, the issue is that a single failure appears to kill the `submit_ray_job()` function, which throws the SIGTERM that kills the ray cluster, which then spins up a new ray cluster. Rinse / lather / repeat.

(continuing...)

**Comment by psschwei** [comment]
Given that a lot of things happen async on Kubernetes, it's probably worth getting some kind of retry logic in and/or revisiting our readiness / liveness probes for ray clusters

**Comment by psschwei** [comment]
> cliffhanger :)

sorry, was a lot to type :smiling_face_with_tear:

**Comment by psschwei** [comment]
Here's an example of the scheduler logs with the fix:

```
Deallocated 0 compute resources.
Using selector: EpollSelector
4 free cluster slots.
EXECUTE JOB
EJ --- no authors resource
Head node http://user-424b7c7c-head-svc:8265/ is not ready yet.
EJ2 -- yes compute resource
++++++++++ let's go!
trying to set up job client x0
Using API server address http://user-424b7c7c-head-svc:8265/.
client setup failed
trying to set up job client x1
Using API server address http://user-424b7c7c-head-svc:8265/.
client setup failed
trying to set up job client x2
Using API server address http://user-424b7c7c-head-svc:8265/.
client setup failed
trying to set up job client x3
Using API server address http://user-424b7c7c-head-svc:8265/.
client setup failed
trying to set up job client x4
Using API server address http://user-424b7c7c-head-svc:8265/.
job submission client setup
trying to load json stuff
++++++++++ artifact got
++++++++++ submitting job
trying job submission
job failed
trying job submission
job failed
trying job submission
Uploading package gcs://_ray_pkg_d4b56dcadc50d4c4.zip.
Creating a file package for local directory '/usr/src/app/media/tmp/8ee3c2a6-c1a6-4e41-8659-557dc46e4bf1'.
Rewrote runtime_env `pip` field from [] to None.
job succeeded!
mark success true
++++++++++ job submitted ++++++++++
Executing <Job d341fb58-5b63-4b84-9724-fa8a288b0dc8 | PENDING>
1 are scheduled for execution.
Using selector: EpollSelector
Using API server address http://user-424b7c7c-head-svc:8265/.
Starting new HTTP connection (1): user-424b7c7c-head-svc:8265
http://user-424b7c7c-head-svc:8265 "GET /api/version HTTP/1.1" 200 98
Starting new HTTP connection (1): user-424b7c7c-head-svc:8265
http://user-424b7c7c-head-svc:8265 "GET /api/version HTTP/1.1" 200 98
Starting new HTTP connection (1): user-424b7c7c-head-svc:8265
http://user-424b7c7c-head-svc:8265 "GET /api/jobs/raysubmit_18BhtX2KKKWRhLKF/logs HTTP/1.1" 200 12
Starting new HTTP connection (1): user-424b7c7c-head-svc:8265
http://user-424b7c7c-head-svc:8265 "GET /api/jobs/raysubmit_18BhtX2KKKWRhLKF HTTP/1.1" 200 900
Job [d341fb58-5b63-4b84-9724-fa8a288b0dc8] status changed from [PENDING] to [RUNNING]
Updated 1 jobs.
Using selector: EpollSelector
Deallocated 0 compute resources.
```

**Comment by IceKhan13** [comment]
@psschwei do we want to add this retry logic here or in a separate PR (not to mix workstreams)?

**Review by psschwei — APPROVED** [review]

**Pull request merged**